packages/headless/src/controllers/core/generated-answer/headless-searchapi-generated-answer.ts

```diff
@@ -0,0 +1,184 @@
+
+      if (
+        subscribeStateManager.engines[genQaEngineId].lastRequestId !== requestId
+      ) {
+        subscribeStateManager.engines[genQaEngineId].lastRequestId = requestId;
+        subscribeStateManager.engines[genQaEngineId].abortController?.abort();
+        engine.dispatch(resetAnswer());
+      }
+
+      const isStreamInProgress =
+        subscribeStateManager.getIsStreamInProgress(genQaEngineId);
+      if (
+        !isStreamInProgress &&
+        streamId &&
+        streamId !== subscribeStateManager.engines[genQaEngineId].lastStreamId
+      ) {
+        subscribeStateManager.engines[genQaEngineId].lastStreamId = streamId;
+        engine.dispatch(
+          streamAnswer({
+            setAbortControllerRef: (ref: AbortController) =>
+              subscribeStateManager.setAbortControllerRef(ref, genQaEngineId),
+          })
+        );
+      }
+    };
+    return engine.subscribe(strictListener);
+  },
+};
+
+export interface SearchAPIGeneratedAnswerAnalyticsClient
+  extends GeneratedAnswerAnalyticsClient {}
+
+/**
+ * Creates a `GeneratedAnswer` controller instance using the search API stream pattern.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `GeneratedAnswer` properties.
+ * @returns A `GeneratedAnswer` controller instance.
+ */
+export function buildSearchAPIGeneratedAnswer(
+  engine: SearchEngine | InsightEngine,
+  analyticsClient: SearchAPIGeneratedAnswerAnalyticsClient,
+  props: GeneratedAnswerProps = {}
+): SearchAPIGeneratedAnswer {
+  if (!loadGeneratedAnswerReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  // const {dispatch} = engine;
+  const controller = buildCoreGeneratedAnswer(engine, analyticsClient, props);
+  const getState = () => engine.state;
+
+  if (!engine.state.generatedAnswer.id) {
+    const genQaEngineId = randomID('genQA-', 12);
+    engine.dispatch(setId({id: genQaEngineId}));
+    subscribeStateManager.engines[genQaEngineId] = {
+      abortController: undefined,
+      lastRequestId: '',
+      lastStreamId: '',
+    };
+  }
+
+  subscribeStateManager.subscribeToSearchRequests(engine);
+
+  const isSearchEngine = (
+    engine: SearchEngine | InsightEngine
+  ): engine is SearchEngine => {
+    return 'executeFirstSearchAfterStandaloneSearchboxRedirect' in engine;
+  };
+
+  return {
+    ...controller,
+
+    get state() {
+      return getState().generatedAnswer;
+    },
+    retry() {
+      if (!isSearchEngine(engine)) {
+        return;
+      }
+      engine.dispatch(
+        executeSearch({
+          legacy: analyticsClient.logRetryGeneratedAnswer(),
+        })
+      );
+    },
+
+    rephrase(responseFormat: GeneratedResponseFormat) {
+      controller.rephrase(responseFormat);
+      if (!isSearchEngine(engine)) {
+        return;
+      }
+      engine.dispatch(
+        executeSearch({
+          legacy: analyticsClient.logRephraseGeneratedAnswer(responseFormat),
+        })
+      );
```

Comment:
same thing for the rephrase action, do we wanna unit-test that?

Reply:
Yes, I did add a whole test for `retry` too. (reply and rephrase are tested for executeSearch now)